### ISSUE TITLE
LIME for multilabel explanations? 

### ISSUE THREAD

**Issue by Hellisotherpeople** [marcotcr/lime#374]
I implemented a multilabel prediction algorithm for NLP text classification. 

Basically, use multilabelbinarizer, binary_crossentropy, and a final sigmoid activation. Since I'm not using a softmax on my output neurons - it's allowed to make a prediction on any class. 

This means that the probabilities in my "predict_proba" function do not sum to 1, and this causes issues with LIME. 

Is there a way around this? Can LIME work for Multilabel classification models?

**Comment by RekhaChandrasekaran** [comment]
I'm also working on a similar problem. As you mentioned, using the multi-label output probabilities with LIME didn't work well. So I wrote a wrapper to explain one output label at a time.
Would be helpful if LIME can support multi-label outputs out of the box.

**Comment by Hellisotherpeople** [comment]
@RekhaChandrasekaran 

Can you post the code for that rapper? I'm willing to try to utilize something like this within my own system.

**Comment by RekhaChandrasekaran** [comment]
Sure. I just wrote a simple wrapper outside of LIME, to explain one output label at a time.
```
def multi_label_explainer(input_text, labels, num_features=None, num_samples=None, bow=True):
    # labels - all outputs
    for label in labels:
        class_names = ['None', label]

        def make_classifier_pipeline(label=label):
            label_index = labels.index(label)
            # pick the corresponding output node 
            def lime_explainer_pipeline(texts):    
                X_train_sequences = [text_to_index_sequence(text_processor.cn_preprocessing_2(text), spacy_nlp) for text in texts]
                X_train_sequences = pad_sequences(X_train_sequences, maxlen=MAX_SEQUENCE_LENGTH, padding='post', truncating='post')
                features = embeddings_matrix[X_train_sequences]
                predict_probs = model.predict(features)
                prob_true = predict_probs[:, label_index]
                result = np.transpose(np.vstack(([1-prob_true, prob_true])))  
                result  = result.reshape(-1, 2)
                #print(predict_probs)
                #print(result)
                return result
        
            return lime_explainer_pipeline
       
       # make a classifier function for the required label
        classifer_fn = make_classifier_pipeline(label=label)

        # ignore this code
        if num_samples is None:
            num_samples = int(len(input_text.split(' ')) * 2.5)
            num_samples = 1000 if num_samples > 1000 else num_samples
        if num_features is None:
            num_features = int(len(input_text.split(' ')) // 20)
            num_features = 10 if num_features > 10 else num_features

        explainer = LimeTextExplainer(class_names=class_names, kernel_width=25, bow=bow)
        exp = explainer.explain_instance(input_text, classifer_fn, num_features=num_features, num_samples=num_samples)
        exp.show_in_notebook(text=True, predict_proba=True)
```

**Comment by Hellisotherpeople** [comment]
One thing that you can do (at least for the closely related ELI5) library is to force your probabilities to sum to 1... e.g. run your computations through a final softmax.

**Comment by marcotcr** [comment]
I think probabilities not summing to one is not a problem. We just display a warning, but it should still work. Please reopen if not : )
